### PR TITLE
Supporting Tensorflow 2

### PR DESCRIPTION
**Pull request by duhaime**
Hello! The folks at @yaledhlab are huge fans of your work @minimaxir. I wanted to send on the first draft of a pull request to make the library function in Tensorflow 1 or 2. 

The only feature not ported to tf2 here is the memory saving gradients logic. The upstream library from which that logic is adopted still hasn't been ported to tf2, and some of the underlying graph traversal methods have been removed from tensorflow itself in 2.x, so there would be a good bit of work getting this running. 

That said, the memory saving gradients are simply a performance enhancement. For interested parties, there's a [thread](https://github.com/cybertronai/gradient-checkpointing/issues/51) going in @cybertronai/gradient-checkpointing and another thread in the @tensorflow repo that discusses some decorater-based approaches to gradient checkpointing in tf2...

In any event, thanks for this great work!

### CODE REVIEW

**Comment by IoIxD** [comment]
Actually I don't think anyone is even maintaining this. The last push was in February and the last issue closure was in March.

**Comment by minimaxir** [comment]
I'm maintaining it but it's less of a priority relative to my other work.

If this does support TF2 then I can investigate.

**Comment by duhaime** [comment]
@minimaxir that's totally understandable! Yes, you should find this will support TF2--I was building TF2 and TF1 models yesterday.

I could create some simple tests for TF1 and TF2, but we'd want some kind of automated environment like Travis/Jenkins where the various Tensorflow versions could be tested...

**Comment by tiagovaz** [comment]
I've tested this PR over TF2.6.0 and everything works nicely so far, thanks @duhaime!

**Comment by L33Tech** [comment]
Unfortunately this no longer seems to work - back_prop seems to be depreciated. I tried to figure out a fix but was unable to find what changes were needed. Hopefully @duhaime can figure this out? Big fan of your work!

**Comment by L33Tech** [comment]
Thanks for the response - I'm running TF 2.6.0 but they seem to have updated & depreciated something sometime in the past week. (Maybe 2.7.0rc0?)
I used to get the following message, but it would proceed anyway:
image

Earlier today it started saying that back_prop was depreciated entirely and failing to produce any output at all.

I've reinstalled and rolled back things in the mean time to stay on an older version (which is why I don't have a screenshot of the new error), but hopefully this can be fixed!
If there's any more info you need, please let me know! Happy to help with this any way I can.

**Comment by L33Tech** [comment]
@duhaime I guess it must be something wrong on my end then, sorry about that. I'd say go ahead with unstacking the last commit and adding support for up to 2.6. Thanks for the awesome work and the help!

**Comment by minimaxir** [comment]
Thanks everyone for your contributions! I have time this week to verify things work; if they do, I'll merge and push out a new release to PyPI.

**Comment by minimaxir** [comment]
Testing it seems fine; working on a 0.8.0 release now!

**Comment by minimaxir** [comment]
v0.8.1 is out, let me know if that works and feel free to file another PR if there's something else broken.
